ta/pkcs11/src/pkcs11_token.c

```diff
@@ -157,19 +163,70 @@ uint32_t entry_ck_slot_info(uint32_t ptypes, TEE_Param *params)
 	if (!token)
 		return PKCS11_CKR_SLOT_ID_INVALID;
 
-	/* Pad string identifiers with blank characters */
-	n = strnlen((char *)info.slot_description,
-		    sizeof(info.slot_description));
-	TEE_MemFill(&info.slot_description[n], ' ',
-		    sizeof(info.slot_description) - n);
-
-	n = strnlen((char *)info.manufacturer_id,
-		    sizeof(info.manufacturer_id));
-	TEE_MemFill(&info.manufacturer_id[n], ' ',
-		    sizeof(info.manufacturer_id) - n);
+	padd_str(info.slot_description, sizeof(info.slot_description));
+	padd_str(info.manufacturer_id, sizeof(info.manufacturer_id));
 
 	out->memref.size = sizeof(info);
 	TEE_MemMove(out->memref.buffer, &info, out->memref.size);
 
 	return PKCS11_CKR_OK;
 }
+
+uint32_t entry_ck_token_info(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = &params[0];
+	TEE_Param *out = &params[2];
+	uint32_t rv = 0;
+	struct serialargs ctrlargs = { };
+	uint32_t token_id = 0;
+	struct ck_token *token = NULL;
+	struct pkcs11_token_info info = {
+		.manufacturer_id = PKCS11_TOKEN_MANUFACTURER,
+		.model = PKCS11_TOKEN_MODEL,
+		.serial_number = PKCS11_TOKEN_SERIAL_NUMBER,
+		.max_session_count = UINT32_MAX,
+		.max_rw_session_count = UINT32_MAX,
+		.max_pin_len = PKCS11_TOKEN_PIN_SIZE,
+		.min_pin_len = 10,
```

Comment:
This seems to be a magic number.

Reply:
:) will change into a macro.

Reply:
Note value 10 is not govern by the spec. Even value 0 looks a valid size but I though a min could be required. Should I remove this extra constraint?

Reply:
Looking back the spec, a token may return specific value (`CKR_PIN_LEN_RANGE`) upon a too short pin length min length is left to the implementation. So I guess we can hard code a `PKCS11_TOKEN_PIN_SIZE_MIN` in TA. I'll default define to 10 until someone complains.